### PR TITLE
Fix handling of zoo in evalf

### LINKED ISSUE #8242: (1/x).evalf(subs={x: 0}) produces 0
Here's how it looks like:

``` python
In [1]: sp.__version__
Out[1]: '0.7.5-git'

In [2]: x = sp.abc.x

In [3]: (1/x).evalf(subs={x: 0})
Out[3]: 0

In [4]: (1/x).subs({x: 0})
Out[4]: zoo

In [5]: (1/x).xreplace({x: 0})
Out[5]: zoo
```

I did investigate it some time ago and, AFAIR, it happens in `Pow` evaluation that short-circuits to 0 whenever the base is 0 while it should return infs if the exponent is negative.

### LINKED ISSUE #21147: Evalf and evaluate=False issues
`Mul(2, Pow(0,-1, evaluate=False), evaluate=False).evalf()` returns 0 instead of complex infinity.

### PR DESCRIPTION
**Pull request by eagleoflqj**
#### References to other Issues or PRs

Fixes #8242
Fixes #21147
Closes #21183
#### Brief description of what is fixed or changed

Fix by allowing `evalf`s to return `zoo` in addition to mpf value tuple, as there's no equivalent mpc representation in mpmath. Check this return value in each `evalf`s and handle it.
#### Other comments

Some additional logics are added to `equals` and `is_constant` in order to fix newly broken tests. One case is leaving unfixed, which imho is acceptable because it’s a bit complicated and previously based on wrong evalf intermediate result.

#### Release Notes

* core
  * Evaluation of expressions involving zoo is now handled properly in evalf without raising obscure exceptions.

### CODE REVIEW

**Comment by eagleoflqj** [comment]
> I don't want to add any type hints until mypy is running in CI

Then I’d wait #22212 to be merged first

File: sympy/core/expr.py [hunk 1]
```diff
@@ -47,6 +47,8 @@ class Expr(Basic, EvalfMixin):
 
     is_scalar = True  # self derivative is 1
 
+    args: tuple
```
Comment (oscarbenjamin) [inline sympy/core/expr.py:50]: It should be a tuple of Basic. See the same hint in `basic.py`. Is it necessary to add this hint for Expr or can it just inherit the hint from Basic?

File: sympy/core/evalf.py [hunk 2]
```diff
@@ -82,7 +93,13 @@ class PrecisionExhausted(ArithmeticError):
 """
 
 
-def fastlog(x):
+MPF_VALUE_TUPLE = tTuple[int, int, int, int]
+SCALED_ZERO_TUPLE = tTuple[List[int], int, int, int]
+TEMP_RESULT = Any
```
Comment (oscarbenjamin) [inline sympy/core/evalf.py:98]: These names are quite long and I'm not sure what they really mean. If we're going to come up with names for things like this then they need to be informative names and there should be some comments here explaining what they mean. The name `TEMP_RESULT` does not convey much information to me and the fact that it is an alias for `Any` means that it gives the appearance of a meaningful type hint when in fact the hint says nothing.

File: sympy/core/evalf.py [hunk 3]
```diff
@@ -82,7 +93,13 @@ class PrecisionExhausted(ArithmeticError):
...
+# tUnion[tTuple[Optional[MPF_VALUE_TUPLE], Optional[MPF_VALUE_TUPLE], Optional[int], Optional[int]], 'ComplexInfinity']
+
+
+def fastlog(x: Optional[MPF_VALUE_TUPLE]) -> tUnion[int, Any]:
```
Comment (oscarbenjamin) [inline sympy/core/evalf.py:102]: Why is this `Optional`?

File: sympy/core/evalf.py [hunk 4]
```diff
@@ -186,13 +210,13 @@ def scaled_zero(mag, sign=1):
         raise ValueError('scaled zero expects int or scaled_zero tuple.')
 
 
-def iszero(mpf, scaled=False):
+def iszero(mpf: tUnion[MPF_VALUE_TUPLE, SCALED_ZERO_TUPLE, None], scaled=False) -> Optional[bool]:
     if not scaled:
         return not mpf or not mpf[1] and not mpf[-1]
     return mpf and isinstance(mpf[0], list) and mpf[1] == mpf[-1] == 1
```
Comment (oscarbenjamin) [inline sympy/core/evalf.py:216]: Is the return type here not just `bool`?

File: sympy/core/evalf.py [hunk 5]
```diff
@@ -205,6 +229,8 @@ def complex_accuracy(result):
     In the worst case (re and im equal), this is wrong by a factor
     sqrt(2), or by log2(sqrt(2)) = 0.5 bit.
     """
+    if not isinstance(result, tuple):
+        return INF
```
Comment (oscarbenjamin) [inline sympy/core/evalf.py:233]: When would this be called with something that isn't a tuple? Maybe that should be fixed at the calling site rather than here.

File: sympy/core/evalf.py [hunk 6]
```diff
@@ -284,10 +315,12 @@ def finalize_complex(re, im, prec):
     return re, im, re_acc, im_acc
 
 
-def chop_parts(value, prec):
+def chop_parts(value: TEMP_RESULT, prec: int) -> TEMP_RESULT:
     """
     Chop off tiny real or complex parts.
     """
+    if not isinstance(value, tuple):
+        return value
```
Comment (oscarbenjamin) [inline sympy/core/evalf.py:323]: When is this called with something that is not a tuple?

File: sympy/core/expr.py (hunk 1 above)
Comment (eagleoflqj) [reply sympy/core/expr.py:50]: Indeed unnecessary. I should instead add type hints to sub classes of `Expr`.

File: sympy/core/evalf.py (hunk 2 above)
Comment (eagleoflqj) [reply sympy/core/evalf.py:98]: For `MPF_VALUE_TUPLE` and `TEMP_RESULT`, I didn't come up with them randomly, they are just below the comment https://github.com/sympy/sympy/blob/9019765edb0d0b4bbd29674a58b93ef4e75b92f8/sympy/core/evalf.py#L60-L83 which already "define" these names.

Comment (eagleoflqj) [reply sympy/core/evalf.py:98]: For `SCALED_ZERO_TUPLE`, it's used in the functions related to `scaled_zero` and can be found at https://github.com/sympy/sympy/blob/9019765edb0d0b4bbd29674a58b93ef4e75b92f8/sympy/core/evalf.py#L152-L156

Comment (eagleoflqj) [reply sympy/core/evalf.py:98]: `TEMP_RESULT` is `Any` because I failed to find a way that makes mypy happy. I originally use the type `tUnion[tTuple[Optional[MPF_VALUE_TUPLE], Optional[MPF_VALUE_TUPLE], Optional[int], Optional[int]], 'ComplexInfinity']` (commented at the following line) and some variants of it, but they couldn't reflect the fact that if the first argument of the tuple is `None` then the third must also be `None`, and sometimes a temp_result is impossible to be `ComplexInfinity` but mypy still requires to check this case. Dealing with these cases neither helps to fix the bug that this PR aims to fix, nor helps understand and refactor the code in evalf.py (at least now).
Moreover, `TEMP_RESULT` provides a way that unifys the return type of the `evalf`s. So imho it's better than nothing or tons of `Any`s here.

File: sympy/core/evalf.py (hunk 3 above)
Comment (eagleoflqj) [reply sympy/core/evalf.py:102]: Because the body of `fastlog` expect this case.
https://github.com/sympy/sympy/blob/9019765edb0d0b4bbd29674a58b93ef4e75b92f8/sympy/core/evalf.py#L117

File: sympy/core/evalf.py (hunk 4 above)
Comment (eagleoflqj) [reply sympy/core/evalf.py:216]: `iszero(None, True)` returns `None`. This may happen in https://github.com/sympy/sympy/blob/9019765edb0d0b4bbd29674a58b93ef4e75b92f8/sympy/core/evalf.py#L554

File: sympy/core/evalf.py (hunk 5 above)
Comment (eagleoflqj) [reply sympy/core/evalf.py:233]: `zoo`, which isn't represented in mpmath. Originally `TEMP_RESULT` is not `Any` and `if not isinstance(result, tuple)` may make my type checker happy. I will change all of this to `if result is S.ComplexInfinity`.

File: sympy/core/evalf.py (hunk 6 above)
Comment (eagleoflqj) [reply sympy/core/evalf.py:323]: same above

File: sympy/core/evalf.py (hunk 2 above)
Comment (oscarbenjamin) [reply sympy/core/evalf.py:106]: If `TEMP_RESULT` is just an alias for `Any` then how is it better than `Any`?

Note that my main point in the comment above is that there should be comments in the code explaining what these types mean and why they are defined in this way.

File: sympy/core/evalf.py (hunk 3 above)
Comment (oscarbenjamin) [reply sympy/core/evalf.py:110]: I guess what I mean is when is this called with `None`?

File: sympy/core/expr.py [hunk 7]
```diff
@@ -752,6 +749,12 @@ def equals(self, other, failing_expression=False):
             # then this can't be a zero
             return False
 
+        factors = diff.as_coeff_mul()[1]
+        if len(factors) > 1:  # avoid infinity recursion
+            fac_zero = [fac.equals(0) for fac in factors]
+            if None not in fac_zero:  # every part can be decided
+                return any(fac_zero)
```
Comment (oscarbenjamin) [inline sympy/core/expr.py:756]: Apart from the type hints is this the main change in this PR?

Comment (oscarbenjamin) [reply sympy/core/expr.py:756]: Note that mixing changes together like this makes the diff much more difficult to review because the important behavioural changes are buried in a sea of unrelated changes.

Comment (eagleoflqj) [reply sympy/core/expr.py:756]: No the main change is in `evalf_pow`.
Sorry about mixing changes together. Originally I only want to fix the `0**-1` evalf bug, then I feel the `evalf.py` is hard to modify without type hints, then the fix makes some tests fail and I have to patch other modules, then `mypy` reports tons of errors...
Since CI passes now, I will re-arrange the commits by logic.

File: sympy/core/evalf.py (hunk 2 above)
Comment (eagleoflqj) [reply sympy/core/evalf.py:106]: > If `TEMP_RESULT` is just an alias for `Any` then how is it better than `Any`?

Because you can not tell 2 `Any`s are actually the same type or not, but alias convey the information that they are the same type.

Comment (eagleoflqj) [reply sympy/core/evalf.py:106]: > Note that my main point in the comment above is that there should be comments in the code explaining what these types mean and why they are defined in this way.

`An mpf value tuple is a tuple of integers (sign, man, exp, bc) ...` clearly explains why `MPF_VALUE_TUPLE = tTuple[int, int, int, int]` and what these 4 `int`s are. In fact I don't quite understand what scaled_zero tuple is, this name already appears in the comment, confusing me, so I simply add type hints for it.

Comment (oscarbenjamin) [reply sympy/core/evalf.py:106]: It's okay to write a comment in the code saying that you don't understand something. That kind of comment is actually very helpful to other readers,

File: sympy/core/evalf.py (hunk 3 above)
Comment (eagleoflqj) [reply sympy/core/evalf.py:110]: https://github.com/sympy/sympy/blob/d04fa46664c3aeda08fb41a2b2d62055056e8926/sympy/core/evalf.py#L880-L881
or
https://github.com/sympy/sympy/blob/d04fa46664c3aeda08fb41a2b2d62055056e8926/sympy/core/evalf.py#L1000-L1006
where the argument of `fastlog` is directly from a call to `evalf`s without checking `None`.

**Comment by oscarbenjamin** [comment]
Can you split this PR into two PRs where one fixes the cited issues and the other adds type hints?

**Comment by eagleoflqj** [comment]
> Can you split this PR into two PRs where one fixes the cited issues and the other adds type hints?

Much clearer now

File: sympy/core/tests/test_evalf.py [hunk 8]
```diff
@@ -628,3 +633,28 @@ def test_issue_20291():
...
+
+
+def test_evalf_with_zoo():
+    assert (1/x).evalf(subs={x: 0}) == zoo  # issue 8242
+    assert (-1/x).evalf(subs={x: 0}) == zoo  # PR 16150
+    assert (0 ** x).evalf(subs={x: -1}) == zoo  # PR 16150
+    assert (0 ** x).evalf(subs={x: -1 + I}) == nan
+    assert Mul(2, Pow(0, -1, evaluate=False), evaluate=False).evalf() == zoo  # issue 21147
+    assert Mul(x, 1/x, evaluate=False).evalf(subs={x: 0}) == Mul(x, 1/x, evaluate=False).subs(x, 0) == nan
+    assert Mul(1/x, 1/x, evaluate=False).evalf(subs={x: 0})
```
Comment (oscarbenjamin) [inline sympy/core/tests/test_evalf.py:645]: There seems to be a missing `== zoo` here

File: sympy/functions/elementary/exponential.py [hunk 9]
```diff
@@ -1055,6 +1055,9 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
             return self.func(x0) - 2*I*S.Pi
         return self.func(arg)
 
+    def is_constant(self, *wrt, **flags):
+        return self.args[0].is_constant(*wrt, **flags)
```
Comment (oscarbenjamin) [inline sympy/functions/elementary/exponential.py:1059]: Why is this needed for `log`? No other function class has this.

**Comment by oscarbenjamin** [comment]
This is much clearer, thanks. I think this looks good although I left a couple of comments.

It seems a bit awkward that everywhere in `evalf` has to check for `zoo` but I can't really think of a better way of handling it.

File: sympy/functions/elementary/exponential.py (hunk 9 above)
Comment (eagleoflqj) [reply sympy/functions/elementary/exponential.py:1059]: It was imported when fixing test_issue_22020 in test_expr.py
Seems without it that case can still pass now. I'll remove it to see what CI says.

**Comment by oscarbenjamin** [comment]
I changed the title of the PR and also edited the release note.

This looks good. Thanks for this!

**Comment by oscarbenjamin** [comment]
I bet this fixes many more issues that aren't explicitly listed in the OP as well.
